[Clang] add emit -Wignored-base-class-qualifiers diagnostic for cv-qualified base classes

lyledean1 opened the pull request:
Split diagnosing base class qualifiers from the ``-Wignored-Qualifiers`` diagnostic group into a new ``-Wignored-base-class-qualifiers`` diagnostic group (which is grouped under ``-Wignored-qualifiers``). 

Fixes: https://github.com/llvm/llvm-project/issues/131935

lyledean1 committed:
Add Ignored CV Qualifiers

lyledean1 committed:
remove build

lyledean1 committed:
Update ReleaseNotes.rst

lyledean1 commented:
@erichkeane , I've added what I think is the correct approach for https://github.com/llvm/llvm-project/issues/131935 by adding a new group to [DiagnosticGroups.td](https://github.com/llvm/llvm-project/pull/132116/files#diff-c21da66b943620c063559b855971be3b2622d5ba1e99e9ff6d9cf0bf132b9fce)

Not sure on what the best name is for the new group, so added "IgnoredCVQualifiers" for the time being 

Let me know if this is going in the right direction - feedback would be appreciated and set as draft for the time being

erichkeane reviewed (commented):
This I think ends up being the 'minimum' patch here, but see my suggestion on the bug:

>I also believe there is value to only diagnosing this on EXPLICIT qualifiers (or a separate, additional warning to do so), not ones that get picked up from the template.

The more I think about it, the more I think that we should modify the warning to only happen on EXPLICIT qualifiers.  So:

```
struct Base{};
using CBase = const Base;

struct D1 : const Base {}; // DOES warn
struct D2 : CBase{}; // a DIFFERENT warning (same wording perhaps?) under a different group.
template<typename T> Templ: const T{};
Templ<Base> t; // Causes same warning as D1.
template<typename T> Templ2 : T{};  
Templ2<CBase> t; // Causes same warning as D2.
```

If you search for where that diagnostic is emitted, you should see that.  You might have to do some debugging to figure out how to differentiate the cases though.

AaronBallman commented:
Please also be sure to add enough information to the PR summary to explain why the changes are needed (the summary becomes the commit message).

erichkeane commented:
> > The more I think about it, the more I think that we should modify the warning to only happen on EXPLICIT qualifiers. So:
> > ```
> > struct Base{};
> > using CBase = const Base;
> > 
> > struct D1 : const Base {}; // DOES warn
> > struct D2 : CBase{}; // a DIFFERENT warning (same wording perhaps?) under a different group.
> > template<typename T> Templ: const T{};
> > Templ<Base> t; // Causes same warning as D1.
> > template<typename T> Templ2 : T{};  
> > Templ2<CBase> t; // Causes same warning as D2.
> > ```
> 
> Adding qualifiers directly in base specifiers is invalid (and all major compilers correctly reject this). See grammar in [[class.derived.general]/1](https://timsong-cpp.github.io/cppwp/n4950/class.derived#general-1). Qualifiers can indirectly come from `decltype`, template parameters, type aliases and so on.

Ah, cool.  I didn't realize we rejected that already. Then I think adding this to a new flag is sufficient, so probably OK after Aaron + My comments are addressed.

lyledean1 committed:
Update to base-class-qualifiers name

lyledean1 committed:
Update to base-class-qualifiers name

lyledean1 committed:
Update to base-class-qualifiers name

lyledean1 committed:
Remove toadd comment

lyledean1 commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -284,6 +284,8 @@ Improvements to Clang's diagnostics
 
 - Improve the ``-Wundefined-func-template`` warning when a function template is not instantiated due to being unreachable in modules.
 
+- Clang now emits a ``-Wignored-cv-qualifiers`` diagnostic when a base class includes cv-qualifiers ([toadd]).
```
on the original PR release note i saw a reference to the feature number i.e (#GH55474) and wasn't sure if i needed to add it here - removed it

lyledean1 commented on clang/include/clang/Basic/DiagnosticGroups.td:
```diff
@@ -497,6 +497,7 @@ def GNUImaginaryConstant : DiagGroup<"gnu-imaginary-constant">;
 def IgnoredGCH : DiagGroup<"ignored-gch">;
 def IgnoredReferenceQualifiers : DiagGroup<"ignored-reference-qualifiers">;
 def IgnoredQualifiers : DiagGroup<"ignored-qualifiers", [IgnoredReferenceQualifiers]>;
+def IgnoredCVQualifiers : DiagGroup<"ignored-cv-qualifiers", [IgnoredReferenceQualifiers]>;
```
I've updated this, so all references will be to `ignored-base-class-qualifiers` now

lyledean1 commented:
I've addressed the comments @erichkeane , so its now `-Wignored-base-class-qualifiers` 

@AaronBallman , let me know if that summary isn't sufficient

erichkeane reviewed (approved):
I'm happy when Aaron is happy with commit message/etc.  THOUGH we should do `Fixes: #<BUG NUMBER>` in it so it auto-closes the bug.

I suspect we'll have to merge this for you, so if you notice CI has passed, feel free to ping us and we'll do it.

MagentaTreehouse commented:
I suggest backporting to the 20 release. Wdyt? (If we decide so, the release notes would go into 20 instead of main)

lyledean1 committed:
Merge branch 'main' into add-ignored-cv-qualifiers

lyledean1 commented:
Happy to adjust this @MagentaTreehouse for LLVM 20 - what would that require?

@erichkeane is this ok to merge (assuming I make the changes for the above - the CI has passed)

lyledean1 committed:
Update clang/docs/ReleaseNotes.rst

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

lyledean1 commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -298,6 +298,8 @@ Improvements to Clang's diagnostics
 
 - Improve the ``-Wundefined-func-template`` warning when a function template is not instantiated due to being unreachable in modules.
 
+- Clang now emits a ``-Wignored-base-class-qualifiers`` diagnostic when a base class includes cv-qualifiers.
```
Thanks, will add that in

lyledean1 commented on clang/include/clang/Basic/DiagnosticGroups.td:
```diff
@@ -497,6 +497,7 @@ def GNUImaginaryConstant : DiagGroup<"gnu-imaginary-constant">;
 def IgnoredGCH : DiagGroup<"ignored-gch">;
 def IgnoredReferenceQualifiers : DiagGroup<"ignored-reference-qualifiers">;
 def IgnoredQualifiers : DiagGroup<"ignored-qualifiers", [IgnoredReferenceQualifiers]>;
+def IgnoredBaseClassQualifiers : DiagGroup<"ignored-base-class-qualifiers", [IgnoredReferenceQualifiers]>;
```
Ah yeah, makes sense, thanks

lyledean1 committed:
Update clang/include/clang/Basic/DiagnosticGroups.td

Co-authored-by: Aaron Ballman <aaron@aaronballman.com>

lyledean1 committed:
Update warn base type qualifier tests

lyledean1 commented on clang/test/SemaCXX/warn-base-type-qualifiers.cpp:
```diff
@@ -1,4 +1,4 @@
-// RUN: %clang_cc1 %s -std=c++11 -Wignored-qualifiers -verify
+// RUN: %clang_cc1 %s -std=c++11 -Wignored-base-class-qualifiers -verify
```
I've updated those tests - so will let the CI run now and confirm everything passes

lyledean1 commented:
Thanks for the review @AaronBallman , I've updated the PR from your comments and the CI has passed - is this good to merge now?

MagentaTreehouse commented:
> Happy to adjust this @MagentaTreehouse for LLVM 20 - what would that require?

@lyledean1 I think the process is that the release notes are left unchanged in this PR; after it is merged we add this to [LLVM 20.X Release](https://github.com/llvm/llvm-project/milestone/26) milestone, open a backport PR by cherry-picking the commit ([Example](https://github.com/llvm/llvm-project/pull/131450)), and add changes to the release notes there.

AaronBallman reviewed (approved):
LGTM!

AaronBallman commented:
Once you resolve the merge conflicts so precommit CI comes back green, I can land the changes

MagentaTreehouse commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -298,6 +298,8 @@ Improvements to Clang's diagnostics
 
 - Improve the ``-Wundefined-func-template`` warning when a function template is not instantiated due to being unreachable in modules.
 
+- Split diagnosing base class qualifiers from the ``-Wignored-Qualifiers`` diagnostic group into a new ``-Wignored-base-class-qualifiers`` diagnostic group (which is grouped under ``-Wignored-qualifiers``). Fixes #GH131935.
```
Should it be "controls" instead of "is grouped under"?

MagentaTreehouse replied:
Ok, previous release notes also used this wording, but I do find it confusing...

MagentaTreehouse commented on clang/docs/ReleaseNotes.rst:
```diff
@@ -298,6 +298,8 @@ Improvements to Clang's diagnostics
 
 - Improve the ``-Wundefined-func-template`` warning when a function template is not instantiated due to being unreachable in modules.
 
+- Split diagnosing base class qualifiers from the ``-Wignored-Qualifiers`` diagnostic group into a new ``-Wignored-base-class-qualifiers`` diagnostic group (which is grouped under ``-Wignored-qualifiers``). Fixes #GH131935.
```
The "is grouped under" relationship is the wrong way around. Should it be "controls"?

erichkeane replied:
I think the wording is sufficient.  No need to change IMO.

erichkeane commented:
It isn't clear what the merge conflicts are (github won't tell me for some reason?) but those need fixing before we can merge this.

lyledean1 committed:
Merge branch 'main' into add-ignored-cv-qualifiers

lyledean1 commented:
> It isn't clear what the merge conflicts are (github won't tell me for some reason?) but those need fixing before we can merge this.

It was on the release notes, that's fixed now

lyledean1 committed:
Update ReleaseNotes.rst

lyledean1 commented:
CI seems to be failing with the message `no such option: --break-system-packages` from `pip` for `[buildkite/github-pull-requests/linux-linux-x64](https://buildkite.com/llvm-project/github-pull-requests/builds/164119#0195f387-c1b7-4bc5-8fb8-cc8049f19033)` and [buildkite/github-pull-requests](https://buildkite.com/llvm-project/github-pull-requests/builds/164119)

But haven't changed anything around this?

lyledean1 committed:
Update ReleaseNotes.rst

AaronBallman commented:
Precommit CI failure on Linux appears to be unrelated, so landing the changes (I'll watch the bots in case it was related somehow).
